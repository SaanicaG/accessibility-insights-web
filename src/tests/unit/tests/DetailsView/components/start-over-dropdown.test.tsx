// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ActionButton, ContextualMenu, IButton, IRefObject } from '@fluentui/react';
import { render } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';

import { InsightsCommandButton } from 'common/components/controls/insights-command-button';
import { StartOverContextMenuKeyOptions } from 'DetailsView/components/details-view-right-panel';
import { StartOverDialogType } from 'DetailsView/components/start-over-dialog';
import * as React from 'react';
import {
    expectMockedComponentPropsToMatchSnapshots,
    getMockComponentClassPropsForCall,
    mockReactComponents,
    useOriginalReactElements,
} from 'tests/unit/mock-helpers/mock-module-helpers';
import { IMock, Mock, Times } from 'typemoq';

import {
    StartOverDropdown,
    StartOverProps,
} from '../../../../../DetailsView/components/start-over-dropdown';
jest.mock('@fluentui/react');
jest.mock('common/components/controls/insights-command-button');

describe('StartOverDropdownTest', () => {
    mockReactComponents([ContextualMenu, InsightsCommandButton]);
    let defaultProps: StartOverProps;
    let openDialogMock: IMock<(dialogType: StartOverDialogType) => void>;

    beforeEach(() => {
        openDialogMock = Mock.ofInstance(() => null);
        defaultProps = {
            singleTestSuffix: 'single test suffix stub',
            allTestSuffix: 'all test suffix stub',
            rightPanelOptions: { showAssessment: true, showTest: true },
            switcherStartOverPreferences: { showAssessment: true, showTest: true },
            dropdownDirection: 'down',
            openDialog: openDialogMock.object,
            buttonRef: {} as IRefObject<IButton>,
        };
    });

    it('render', () => {
        const renderResult = render(<StartOverDropdown {...defaultProps} />);

        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([ContextualMenu, InsightsCommandButton]);
    });

    it('render ContextualMenu', async () => {
        useOriginalReactElements('common/components/controls/insights-command-button', [
            'InsightsCommandButton',
        ]);

        const renderResult = render(<StartOverDropdown {...defaultProps} />);

        await userEvent.click(renderResult.getByRole('button'));

        expect(renderResult.asFragment()).toMatchSnapshot();

        const mockProps = getMockComponentClassPropsForCall(ContextualMenu);
        expect(mockProps.target).not.toBeNull();
    });

    const menuButtonOptions = [true, false];
    const optionName = 'showTest';
    const optionKey = 'test';
    menuButtonOptions.forEach(rightPanelOptionEnabled => {
        menuButtonOptions.forEach(switcherPreferencesOptionEnabled => {
            const rightPanelOptions = {
                [optionName]: rightPanelOptionEnabled,
            } as StartOverContextMenuKeyOptions;
            const switcherPreferences = {
                [optionName]: switcherPreferencesOptionEnabled,
            } as StartOverContextMenuKeyOptions;

            const shouldFindOption = rightPanelOptionEnabled && switcherPreferencesOptionEnabled;

            const casePrefix = shouldFindOption
                ? `${optionKey} item IS rendered`
                : `${optionKey} item IS NOT rendered`;

            test(`${casePrefix} - rightPanelOptions.${optionName} is ${rightPanelOptionEnabled} & switcherStartOverPreferences.${optionName} is ${switcherPreferencesOptionEnabled}`, async () => {
                defaultProps.rightPanelOptions = rightPanelOptions;
                defaultProps.switcherStartOverPreferences = switcherPreferences;

                const renderResult = render(<StartOverDropdown {...defaultProps} />);
                await userEvent.click(renderResult.getByRole('button'));
                const isStartOverOptionRendered = getMockComponentClassPropsForCall(
                    ContextualMenu,
                ).items.some(item => item.key === optionKey);
                expect(isStartOverOptionRendered).toEqual(shouldFindOption);
            });
        });
    });

    it('render with dropdown on left', async () => {
        const props: StartOverProps = {
            ...defaultProps,
            dropdownDirection: 'left',
        };

        const renderResult = render(<StartOverDropdown {...props} />);
        await userEvent.click(renderResult.getByRole('button'));
        expect(renderResult.asFragment()).toMatchSnapshot();
    });

    it('should open the start test over dialog', async () => {
        openDialogMock.setup(sds => sds('test')).verifiable(Times.once());

        const renderResult = render(<StartOverDropdown {...defaultProps} />);
        await userEvent.click(renderResult.getByRole('button'));
        getMockComponentClassPropsForCall(ContextualMenu)
            .items.find(elem => elem.key === 'test')
            .onClick();

        openDialogMock.verifyAll();
    });

    it('should open the start assessment over dialog', async () => {
        openDialogMock.setup(sds => sds('assessment')).verifiable(Times.once());

        const renderResult = render(<StartOverDropdown {...defaultProps} />);
        await userEvent.click(renderResult.getByRole('button'));
        getMockComponentClassPropsForCall(ContextualMenu)
            .items.find(elem => elem.key === 'assessment')
            .onClick();

        openDialogMock.verifyAll();
    });

    it('should dismiss the contextMenu', async () => {
        const renderResult = render(<StartOverDropdown {...defaultProps} />);

        await userEvent.click(renderResult.getByRole('button'));
        getMockComponentClassPropsForCall(ContextualMenu).onDismiss();
        const mockContextualmenu = renderResult.container.querySelector('mock-contextual-menu');
        expect(mockContextualmenu).toBeNull();
    });
});
