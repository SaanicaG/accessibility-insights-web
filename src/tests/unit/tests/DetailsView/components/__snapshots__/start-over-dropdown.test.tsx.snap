// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartOverDropdownTest render 1`] = `
<DocumentFragment>
  <div>
    <mock-insightscommandbutton
      arialabel="start over menu"
      componentref="[object Object]"
      iconprops="[object Object]"
      menuiconprops="[object Object]"
      text="Start over"
    />
  </div>
</DocumentFragment>
`;

exports[`StartOverDropdownTest render ContextualMenu 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-label="start over menu"
      class="ms-Button ms-Button--action ms-Button--command root-109"
      data-is-focusable="true"
      type="button"
    >
      <span
        class="ms-Button-flexContainer flexContainer-110"
        data-automationid="splitbuttonprimary"
      >
        <i
          aria-hidden="true"
          class="ms-Icon root-105 ms-Button-icon icon-112"
          data-icon-name="Refresh"
        />
        <span
          class="ms-Button-textContainer textContainer-111"
        >
          <span
            class="ms-Button-label label-113"
            id="id__0"
          >
            Start over
          </span>
        </span>
        <i
          aria-hidden="true"
          class="ms-Icon root-105 ms-Button-menuIcon menuIcon-114"
          data-icon-name="ChevronDown"
        />
      </span>
    </button>
    <span
      class="ms-layer"
    />
  </div>
</DocumentFragment>
`;

exports[`StartOverDropdownTest render with dropdown on left 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-label="start over menu"
      class="ms-Button ms-Button--action ms-Button--command root-109"
      data-is-focusable="true"
      type="button"
    >
      <span
        class="ms-Button-flexContainer flexContainer-110"
        data-automationid="splitbuttonprimary"
      >
        <i
          aria-hidden="true"
          class="ms-Icon root-105 ms-Button-icon icon-112"
          data-icon-name="Refresh"
        />
        <span
          class="ms-Button-textContainer textContainer-111"
        >
          <span
            class="ms-Button-label label-113"
            id="id__30"
          >
            Start over
          </span>
        </span>
        <i
          aria-hidden="true"
          class="ms-Icon root-105 ms-Button-menuIcon menuIcon-114"
          data-icon-name="ChevronRight"
        />
      </span>
    </button>
    <span
      class="ms-layer"
    />
  </div>
</DocumentFragment>
`;

exports[`StartOverDropdownTest render: ContextualMenu props 1`] = `undefined`;

exports[`StartOverDropdownTest render: InsightsCommandButton props 1`] = `
{
  "ariaLabel": "start over menu",
  "componentRef": {},
  "iconProps": {
    "iconName": "Refresh",
  },
  "menuIconProps": {
    "iconName": "ChevronDown",
  },
  "onClick": [Function],
  "text": "Start over",
}
`;

exports[`StartOverDropdownTest render: action button menu props 1`] = `undefined`;
